File: engine-go/internal/postgres/parametrico.go
package postgres

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/provincia/palestre-engine/internal/calc"
	"github.com/provincia/palestre-engine/internal/istruttoria"
	"github.com/provincia/palestre-engine/internal/roundrobin"
	"github.com/shopspring/decimal"
)

// ParametricoCaricato raccoglie la versione parametrica attiva (l'ultima pubblicata per
// valida_dal, vedi CLAUDE.md) e le tabelle di scaglioni normative, pronte per
// calc/istruttoria/roundrobin.
type ParametricoCaricato struct {
	VersioneID         string
	Istruttoria        istruttoria.Parametrico
	Limiti             roundrobin.LimitiConcentrazione
	TolleranzaISF      decimal.Decimal
	PesoFasciaPregiata decimal.Decimal
}

func decimalDaTesto(s string) (decimal.Decimal, error) {
	d, err := decimal.NewFromString(s)
	if err != nil {
		return decimal.Decimal{}, fmt.Errorf("valore decimal non valido %q: %w", s, err)
	}
	return d, nil
}

func decimalDaTestoNullable(s *string) (decimal.Decimal, bool, error) {
	if s == nil {
		return decimal.Decimal{}, false, nil
	}
	d, err := decimalDaTesto(*s)
	if err != nil {
		return decimal.Decimal{}, false, err
	}
	return d, true, nil
}

// CaricaParametricoAttivo carica l'ultima versione parametrica pubblicata e tutte le
// tabelle di scaglioni collegate (versionate e normative globali).
func CaricaParametricoAttivo(ctx context.Context, pool *pgxpool.Pool) (ParametricoCaricato, error) {
	var (
		versioneID                                                                  string
		moltiplicatoreTxt, pesoFasciaTxt, caaNeutroTxt, csdNeutroTxt, tolleranzaTxt string
		minutiSettimanaliMax, slotMax, fascePregiateMax, giornateGaraMax            int
	)
	err := pool.QueryRow(ctx, `
		SELECT id,
		       moltiplicatore_minuti_per_punto::text,
		       peso_fascia_pregiata::text,
		       minuti_settimanali_max::integer,
		       slot_max_stesso_impianto,
		       fasce_pregiate_max,
		       giornate_gara_max,
		       caa_neutro::text,
		       csd_neutro::text,
		       tolleranza_isf_pct::text
		FROM parametrico_versioni
		ORDER BY valida_dal DESC
		LIMIT 1
	`).Scan(&versioneID, &moltiplicatoreTxt, &pesoFasciaTxt, &minutiSettimanaliMax,
		&slotMax, &fascePregiateMax, &giornateGaraMax, &caaNeutroTxt, &csdNeutroTxt, &tolleranzaTxt)
	if err != nil {
		return ParametricoCaricato{}, fmt.Errorf("caricamento parametrico attivo: %w", err)
	}

	moltiplicatore, err := decimalDaTesto(moltiplicatoreTxt)
	if err != nil {
		return ParametricoCaricato{}, err
	}
	pesoFascia, err := decimalDaTesto(pesoFasciaTxt)
	if err != nil {
		return ParametricoCaricato{}, err
	}
	caaNeutro, err := decimalDaTesto(caaNeutroTxt)
	if err != nil {
		return ParametricoCaricato{}, err
	}
	csdNeutro, err := decimalDaTesto(csdNeutroTxt)
	if err != nil {
		return ParametricoCaricato{}, err
	}
	tolleranza, err := decimalDaTesto(tolleranzaTxt)
	if err != nil {
		return ParametricoCaricato{}, err
	}

	incrementoScaglioni, err := caricaIncrementoSquadreScaglioni(ctx, pool)
	if err != nil {
		return ParametricoCaricato{}, err
	}
	crsScaglioni, err := caricaCRSScaglioni(ctx, pool)
	if err != nil {
		return ParametricoCaricato{}, err
	}
	caaScaglioni, err := caricaCAAScaglioni(ctx, pool)
	if err != nil {
		return ParametricoCaricato{}, err
	}
	csdScaglioni, err := caricaCSDScaglioni(ctx, pool, versioneID)
	if err != nil {
		return ParametricoCaricato{}, err
	}

	return ParametricoCaricato{
		VersioneID: versioneID,
		Istruttoria: istruttoria.Parametrico{
			MoltiplicatoreMinutiPerPunto: moltiplicatore,
			IncrementoSquadreScaglioni:   incrementoScaglioni,
			CRSScaglioni:                 crsScaglioni,
			CAAScaglioni:                 caaScaglioni,
			CSDScaglioni:                 csdScaglioni,
			CAANeutro:                    caaNeutro,
			CSDNeutro:                    csdNeutro,
		},
		Limiti: roundrobin.LimitiConcentrazione{
			MinutiSettimanaliMax:  minutiSettimanaliMax,
			SlotMaxStessoImpianto: slotMax,
			FascePregiateMax:      fascePregiateMax,
			GiornateGaraMax:       giornateGaraMax,
		},
		TolleranzaISF:      tolleranza,
		PesoFasciaPregiata: pesoFascia,
	}, nil
}

func caricaIncrementoSquadreScaglioni(ctx context.Context, pool *pgxpool.Pool) ([]calc.ScaglioneSquadre, error) {
	rows, err := pool.Query(ctx, `SELECT squadre_min, squadre_max, incremento FROM incremento_squadre_scaglioni ORDER BY squadre_min`)
	if err != nil {
		return nil, fmt.Errorf("scaglioni incremento squadre: %w", err)
	}
	defer rows.Close()

	var out []calc.ScaglioneSquadre
	for rows.Next() {
		var min, incr int
		var max *int
		if err := rows.Scan(&min, &max, &incr); err != nil {
			return nil, fmt.Errorf("scaglioni incremento squadre: %w", err)
		}
		out = append(out, calc.ScaglioneSquadre{SquadreMin: min, SquadreMax: max, Incremento: incr})
	}
	return out, rows.Err()
}

func caricaCRSScaglioni(ctx context.Context, pool *pgxpool.Pool) ([]calc.ScaglioneCRS, error) {
	rows, err := pool.Query(ctx, `SELECT classe_attivita_codice, livello, crs::text FROM crs_scaglioni`)
	if err != nil {
		return nil, fmt.Errorf("scaglioni CRS: %w", err)
	}
	defer rows.Close()

	var out []calc.ScaglioneCRS
	for rows.Next() {
		var classe string
		var livello *string
		var crsTxt string
		if err := rows.Scan(&classe, &livello, &crsTxt); err != nil {
			return nil, fmt.Errorf("scaglioni CRS: %w", err)
		}
		crs, err := decimalDaTesto(crsTxt)
		if err != nil {
			return nil, err
		}
		out = append(out, calc.ScaglioneCRS{ClasseCodice: classe, Livello: livello, CRS: crs})
	}
	return out, rows.Err()
}

func caricaCAAScaglioni(ctx context.Context, pool *pgxpool.Pool) ([]calc.ScaglioneCAA, error) {
	rows, err := pool.Query(ctx, `SELECT anni_min::text, anni_max::text, caa::text FROM caa_scaglioni ORDER BY anni_min`)
	if err != nil {
		return nil, fmt.Errorf("scaglioni CAA: %w", err)
	}
	defer rows.Close()

	var out []calc.ScaglioneCAA
	for rows.Next() {
		var anniMinTxt string
		var anniMaxTxt *string
		var caaTxt string
		if err := rows.Scan(&anniMinTxt, &anniMaxTxt, &caaTxt); err != nil {
			return nil, fmt.Errorf("scaglioni CAA: %w", err)
		}
		anniMin, err := decimalDaTesto(anniMinTxt)
		if err != nil {
			return nil, err
		}
		anniMax, haMax, err := decimalDaTestoNullable(anniMaxTxt)
		if err != nil {
			return nil, err
		}
		caa, err := decimalDaTesto(caaTxt)
		if err != nil {
			return nil, err
		}
		s := calc.ScaglioneCAA{AnniMin: anniMin, CAA: caa}
		if haMax {
			s.AnniMax = &anniMax
		}
		out = append(out, s)
	}
	return out, rows.Err()
}

func caricaCSDScaglioni(ctx context.Context, pool *pgxpool.Pool, versioneID string) ([]calc.ScaglioneCSD, error) {
	rows, err := pool.Query(ctx, `
		SELECT rapporto_fd_fr_min::text, rapporto_fd_fr_max::text, coefficiente::text
		FROM csd_scaglioni
		WHERE parametrico_versione_id = $1
		ORDER BY rapporto_fd_fr_min
	`, versioneID)
	if err != nil {
		return nil, fmt.Errorf("scaglioni CSD: %w", err)
	}
	defer rows.Close()

	var out []calc.ScaglioneCSD
	for rows.Next() {
		var rapportoMinTxt string
		var rapportoMaxTxt *string
		var coefficienteTxt string
		if err := rows.Scan(&rapportoMinTxt, &rapportoMaxTxt, &coefficienteTxt); err != nil {
			return nil, fmt.Errorf("scaglioni CSD: %w", err)
		}
		rapportoMin, err := decimalDaTesto(rapportoMinTxt)
		if err != nil {
			return nil, err
		}
		rapportoMax, haMax, err := decimalDaTestoNullable(rapportoMaxTxt)
		if err != nil {
			return nil, err
		}
		coefficiente, err := decimalDaTesto(coefficienteTxt)
		if err != nil {
			return nil, err
		}
		s := calc.ScaglioneCSD{RapportoMin: rapportoMin, Coefficiente: coefficiente}
		if haMax {
			s.RapportoMax = &rapportoMax
		}
		out = append(out, s)
	}
	return out, rows.Err()
}

File: backend-node/src/repository/parametrico.ts
import type { Db } from '../db.ts';

export interface ScaglioneCsd {
  rapportoFdFrMin: string;
  rapportoFdFrMax: string | null;
  coefficiente: string;
}

export interface VersioneParametrica {
  id: string;
  validaDal: string;
  pubblicataDa: string | null;
  note: string | null;
  moltiplicatoreMinutiPerPunto: string;
  pesoFasciaPregiata: string;
  minutiSettimanaliMax: string;
  slotMaxStessoImpianto: number;
  fascePregiateMax: number;
  giornateGaraMax: number;
  incrementoSquadreNeutro: number;
  caaNeutro: string;
  csdNeutro: string;
  tolleranzaIsfPct: string;
  sogliaMancatiUtilizziDiffida: number;
  sogliaMancatiUtilizziDecadenza: number;
  sogliaScostamentoDichiaratoPct: string;
  sogliaIsfCompensazione: string;
  retentionLogOperazioniGiorni: number;
  quotaNuoveAssociazioniPct: string;
  termineGiustificazioneGiorni: number;
  creataIl: string;
  csdScaglioni: ScaglioneCsd[];
}

export interface VersioneParametricaSintetica {
  id: string;
  validaDal: string;
  pubblicataDa: string | null;
  note: string | null;
}

interface RigaVersione {
  id: string;
  valida_dal: Date;
  pubblicata_da: string | null;
  note: string | null;
  moltiplicatore_minuti_per_punto: string;
  peso_fascia_pregiata: string;
  minuti_settimanali_max: string;
  slot_max_stesso_impianto: number;
  fasce_pregiate_max: number;
  giornate_gara_max: number;
  incremento_squadre_neutro: number;
  caa_neutro: string;
  csd_neutro: string;
  tolleranza_isf_pct: string;
  soglia_mancati_utilizzi_diffida: number;
  soglia_mancati_utilizzi_decadenza: number;
  soglia_scostamento_dichiarato_pct: string;
  soglia_isf_compensazione: string;
  retention_log_operazioni_giorni: number;
  quota_nuove_associazioni_pct: string;
  termine_giustificazione_giorni: number;
  creata_il: Date;
}

interface RigaScaglioneCsd {
  rapporto_fd_fr_min: string;
  rapporto_fd_fr_max: string | null;
  coefficiente: string;
}

const COLONNE_SELECT_VERSIONE = `id, valida_dal, pubblicata_da, note,
  moltiplicatore_minuti_per_punto::text, peso_fascia_pregiata::text, minuti_settimanali_max::text,
  slot_max_stesso_impianto, fasce_pregiate_max, giornate_gara_max, incremento_squadre_neutro,
  caa_neutro::text, csd_neutro::text, tolleranza_isf_pct::text,
  soglia_mancati_utilizzi_diffida, soglia_mancati_utilizzi_decadenza,
  soglia_scostamento_dichiarato_pct::text, soglia_isf_compensazione::text,
  retention_log_operazioni_giorni, quota_nuove_associazioni_pct::text, termine_giustificazione_giorni, creata_il`;

function daRigaVersione(r: RigaVersione, csdScaglioni: ScaglioneCsd[]): VersioneParametrica {
  return {
    id: r.id,
    validaDal: r.valida_dal.toISOString(),
    pubblicataDa: r.pubblicata_da,
    note: r.note,
    moltiplicatoreMinutiPerPunto: r.moltiplicatore_minuti_per_punto,
    pesoFasciaPregiata: r.peso_fascia_pregiata,
    minutiSettimanaliMax: r.minuti_settimanali_max,
    slotMaxStessoImpianto: r.slot_max_stesso_impianto,
    fascePregiateMax: r.fasce_pregiate_max,
    giornateGaraMax: r.giornate_gara_max,
    incrementoSquadreNeutro: r.incremento_squadre_neutro,
    caaNeutro: r.caa_neutro,
    csdNeutro: r.csd_neutro,
    tolleranzaIsfPct: r.tolleranza_isf_pct,
    sogliaMancatiUtilizziDiffida: r.soglia_mancati_utilizzi_diffida,
    sogliaMancatiUtilizziDecadenza: r.soglia_mancati_utilizzi_decadenza,
    sogliaScostamentoDichiaratoPct: r.soglia_scostamento_dichiarato_pct,
    sogliaIsfCompensazione: r.soglia_isf_compensazione,
    retentionLogOperazioniGiorni: r.retention_log_operazioni_giorni,
    quotaNuoveAssociazioniPct: r.quota_nuove_associazioni_pct,
    termineGiustificazioneGiorni: r.termine_giustificazione_giorni,
    creataIl: r.creata_il.toISOString(),
    csdScaglioni,
  };
}

function daRigaScaglione(r: RigaScaglioneCsd): ScaglioneCsd {
  return { rapportoFdFrMin: r.rapporto_fd_fr_min, rapportoFdFrMax: r.rapporto_fd_fr_max, coefficiente: r.coefficiente };
}

async function caricaScaglioniPerVersione(db: Db, versioneId: string): Promise<ScaglioneCsd[]> {
  const r = await db.query<RigaScaglioneCsd>(
    `SELECT rapporto_fd_fr_min::text, rapporto_fd_fr_max::text, coefficiente::text
     FROM csd_scaglioni WHERE parametrico_versione_id = $1 ORDER BY rapporto_fd_fr_min`,
    [versioneId],
  );
  return r.rows.map(daRigaScaglione);
}

export async function leggiVersioneAttiva(db: Db): Promise<VersioneParametrica | null> {
  const r = await db.query<RigaVersione>(
    `SELECT ${COLONNE_SELECT_VERSIONE} FROM parametrico_versioni ORDER BY valida_dal DESC LIMIT 1`,
  );
  const riga = r.rows[0];
  if (!riga) {
    return null;
  }
  return daRigaVersione(riga, await caricaScaglioniPerVersione(db, riga.id));
}

export async function leggiVersionePerId(db: Db, id: string): Promise<VersioneParametrica | null> {
  const r = await db.query<RigaVersione>(`SELECT ${COLONNE_SELECT_VERSIONE} FROM parametrico_versioni WHERE id = $1`, [id]);
  const riga = r.rows[0];
  if (!riga) {
    return null;
  }
  return daRigaVersione(riga, await caricaScaglioniPerVersione(db, riga.id));
}

export async function listaVersioni(db: Db): Promise<VersioneParametricaSintetica[]> {
  const r = await db.query<{ id: string; valida_dal: Date; pubblicata_da: string | null; note: string | null }>(
    `SELECT id, valida_dal, pubblicata_da, note FROM parametrico_versioni ORDER BY valida_dal DESC`,
  );
  return r.rows.map((riga) => ({
    id: riga.id,
    validaDal: riga.valida_dal.toISOString(),
    pubblicataDa: riga.pubblicata_da,
    note: riga.note,
  }));
}

export interface DatiCreaVersione {
  note?: string | undefined;
  moltiplicatoreMinutiPerPunto: string;
  pesoFasciaPregiata: string;
  minutiSettimanaliMax: string;
  slotMaxStessoImpianto: number;
  fascePregiateMax: number;
  giornateGaraMax: number;
  incrementoSquadreNeutro: number;
  caaNeutro: string;
  csdNeutro: string;
  tolleranzaIsfPct: string;
  sogliaMancatiUtilizziDiffida: number;
  sogliaMancatiUtilizziDecadenza: number;
  sogliaScostamentoDichiaratoPct: string;
  sogliaIsfCompensazione: string;
  retentionLogOperazioniGiorni: number;
  quotaNuoveAssociazioniPct: string;
  termineGiustificazioneGiorni: number;
  csdScaglioni: Array<{ rapportoFdFrMin: string; rapportoFdFrMax: string | null; coefficiente: string }>;
}

export async function creaVersione(
  db: Db,
  dati: DatiCreaVersione,
  pubblicataDa: string | null,
): Promise<VersioneParametrica> {
  const r = await db.query<RigaVersione>(
    `INSERT INTO parametrico_versioni
       (pubblicata_da, note, moltiplicatore_minuti_per_punto, peso_fascia_pregiata, minuti_settimanali_max,
        slot_max_stesso_impianto, fasce_pregiate_max, giornate_gara_max, incremento_squadre_neutro,
        caa_neutro, csd_neutro, tolleranza_isf_pct, soglia_mancati_utilizzi_diffida,
        soglia_mancati_utilizzi_decadenza, soglia_scostamento_dichiarato_pct, soglia_isf_compensazione,
        retention_log_operazioni_giorni, quota_nuove_associazioni_pct, termine_giustificazione_giorni)
     VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19)
     RETURNING ${COLONNE_SELECT_VERSIONE}`,
    [
      pubblicataDa,
      dati.note ?? null,
      dati.moltiplicatoreMinutiPerPunto,
      dati.pesoFasciaPregiata,
      dati.minutiSettimanaliMax,
      dati.slotMaxStessoImpianto,
      dati.fascePregiateMax,
      dati.giornateGaraMax,
      dati.incrementoSquadreNeutro,
      dati.caaNeutro,
      dati.csdNeutro,
      dati.tolleranzaIsfPct,
      dati.sogliaMancatiUtilizziDiffida,
      dati.sogliaMancatiUtilizziDecadenza,
      dati.sogliaScostamentoDichiaratoPct,
      dati.sogliaIsfCompensazione,
      dati.retentionLogOperazioniGiorni,
      dati.quotaNuoveAssociazioniPct,
      dati.termineGiustificazioneGiorni,
    ],
  );
  const versione = r.rows[0]!;
  const scaglioni: ScaglioneCsd[] = [];
  for (const s of dati.csdScaglioni) {
    const rs = await db.query<RigaScaglioneCsd>(
      `INSERT INTO csd_scaglioni (parametrico_versione_id, rapporto_fd_fr_min, rapporto_fd_fr_max, coefficiente)
       VALUES ($1,$2,$3,$4)
       RETURNING rapporto_fd_fr_min::text, rapporto_fd_fr_max::text, coefficiente::text`,
      [versione.id, s.rapportoFdFrMin, s.rapportoFdFrMax, s.coefficiente],
    );
    scaglioni.push(daRigaScaglione(rs.rows[0]!));
  }
  return daRigaVersione(versione, scaglioni);
}
